src/artifacts.js

```diff
@@ -423,6 +643,119 @@ var replyWithArtifact = async function(taskId, runId, name, req, res) {
   this.monitor.reportError(err);
 };
 
+/** Complete artifact */
+api.declare({
+  method:     'put',
+  route:      '/task/:taskId/runs/:runId/artifacts/:name(*)',
+  name:       'completeArtifact',
+  stability:  API.stability.experimental,
+  scopes: [
+    [
+      'queue:create-artifact:<name>',
+      'assume:worker-id:<workerGroup>/<workerId>',
+    ], [
+      'queue:create-artifact:<taskId>/<runId>',
+    ],
+  ],
+  deferAuth:  true,
+  input:      'put-artifact-request.json#',
+  output:     'put-artifact-response.json#',
+  title:      'Complete Artifact',
+  description: 'tbd',
+}, async function(req, res) {
+  let taskId      = req.params.taskId;
+  let runId       = parseInt(req.params.runId, 10);
+  let name        = req.params.name;
+  let input       = req.body;
+  let isPublic    = /^public\//.test(name);
+
+  // Hmm, list destructuring is probably better here
+  let [artifact, task] = await Promise.all([
+    this.Artifact.load({taskId, runId, name}, true),
+    this.Task.load({taskId}, true),
+  ]);
+
+  // TODO: Have a shortcut here where things which are already
+  // marked as present short circuit
+
+  let run = task.runs[runId];
+  if (!run) {
+    return res.reportError('InputError',
+      'Run not found',
+      {});
+  }
+  let workerGroup = run.workerGroup;
+  let workerId = run.workerId;
+
+  // Authenticate request by providing parameters
+  if (!req.satisfies({
+    taskId,
+    runId,
+    workerGroup,
+    workerId,
+    name,
+  })) {
+    return;
+  }
+
+  if (artifact.storageType === 'blob') {
+    let etag;
+    if (artifact.details.partsHash) {
+      etag = await this.s3Controller.completeMultipartUpload({
+        bucket: artifact.details.bucket,
+        key: artifact.details.key,
+        etags: input.etags,
+        uploadId: artifact.details.uploadId,
+        tags: {taskId, runId, name},
+      });
+    } else {
+      let url = await this.s3Controller.generateGetUrl({
+        bucket: artifact.details.bucket,
+        key: artifact.details.key,
+        signed: !isPublic,
+      });
+
+      if (urllib.parse(url).protocol !== 'https:') {
+        throw new Error('Only HTTPS is allowed for artifacts');
+      }
+
+      let headRes = await this.s3Runner.run({
+        req:{
+          url,
+          method: 'HEAD',
+          headers: {},
+        },
+      });
+      
+      if (headRes.headers['x-amz-meta-content-sha256'] !== artifact.details.contentSha256) {
+        throw new Error('S3 object does not have correct Content-Sha256 metadata');
+      }
+      
+      etag = input.etags[0];
+    }
+
+    await artifact.modify((artifact) => {
+      artifact.details.etag = etag;
+      // Now that we're finished, we don't want to store the uploadId any longer
+      artifact.details = _.omit(artifact.details, 'uploadId');
+      artifact.present = true;
+    });
+
+    await this.publisher.artifactCreated({
```

Comment:
Publish before you modify... If publication fails and client retries (as required), the next request won't publish again, because you'll want to return early when `artifact.present == true`

Reply:
What if publication fails 5 times and the retry limit is 5, that means that we'd send a message that the artifact is created, but we wouldn't have modified the artifact so the values in the entity would be all wrong, and we wouldn't hit the shortcircuiting logic if it is there.

That seems like a bad idea to me.

Reply:
@jhford, I have an email that says:
> What if publication fails 5 times and the retry limit is 5, that means that we'd send a message that the artifact is created, but we wouldn't have modified the artifact so the values in the entity would be all wrong, and we wouldn't hit the shortcircuiting logic if it is there.
That seems like a bad idea to me.

You have a point that we return `404` until `present = true` so we don't want to send a message that the artifact is done until such time that `present = true`... Fair argument...

You can fix this by changing the comment that says `// TODO: Return early if present = true` to say that we **cannot** return early, without sending the `artifactCreated` message again...
Either make the code, or make the comment very clear...Otherwise we can't ensure that messages arrive...